File: lib/pmsg/pmsg.dart
import 'package:protobuf/protobuf.dart';
import 'package:protod/google/protobuf/any.pb.dart';
import 'package:protod/pmsg/pmsg.pb.dart';

export 'pmsg.pb.dart';

extension Extensions on Bundle {
  void set(GeneratedMessage m) {
    this.messages[m.info_.qualifiedMessageName] = Any.pack(m);
  }

  bool has(GeneratedMessage m) {
    return this.messages.containsKey(m.info_.qualifiedMessageName);
  }

  T get<T extends GeneratedMessage>(T m) {
    if (!this.has(m)) {
      throw Exception("not found");
    }
    m = m.info_.createEmptyInstance() as T;
    this.messages[m.info_.qualifiedMessageName].unpackInto(m);
    return m;
  }
}

File: pmsg/pmsg.go
package pmsg

import (
	"github.com/dave/protod/delta"
	"github.com/dave/protod/perr"
	any "github.com/golang/protobuf/ptypes/any"
	"google.golang.org/protobuf/proto"
)

func New() *Bundle {
	return &Bundle{Messages: map[string]*any.Any{}}
}

func (b *Bundle) Set(m proto.Message) error {
	a, err := delta.MarshalAny(m)
	if err != nil {
		return perr.Wrap(err).Debug("marshaling message")
	}
	b.Messages[string(m.ProtoReflect().Descriptor().FullName())] = a
	return nil
}

func (b *Bundle) Has(m proto.Message) bool {
	_, found := b.Messages[string(m.ProtoReflect().Descriptor().FullName())]
	return found
}

func (b *Bundle) Get(m proto.Message) (bool, error) {
	a, found := b.Messages[string(m.ProtoReflect().Descriptor().FullName())]
	if !found {
		return false, nil
	}
	if err := delta.UnmarshalAnyInto(a, m); err != nil {
		return false, perr.Wrap(err).Debug("unmarshaling message")
	}
	return true, nil
}

func (b *Bundle) MustSet(m proto.Message) {
	err := b.Set(m)
	if err != nil {
		panic(err)
	}
}

func (b *Bundle) MustGet(m proto.Message) {
	_, err := b.Get(m)
	if err != nil {
		panic(err)
	}
}
